universal-login-contracts/test/contracts/ERC1077.js

```diff
@@ -0,0 +1,303 @@
+  let signature;
+  let msg;
+  let mockContract;
+  let wallet;
+  let mockToken;
+  let anotherWallet;
+  let invalidSignature;
+  let relayerBalance;
+  let relayerTokenBalance;
+
+  beforeEach(async () => {
+    ({provider, identity, privateKey, keyAsAddress, publicKey, mockToken, mockContract, wallet} = await testHelper.load(basicIdentity));
+    msg = {...transferMessage, from: identity.address};
+    signature = calculateMessageSignature(privateKey, msg);
+    [anotherWallet] = await getWallets(provider);
+    invalidSignature = calculateMessageSignature(anotherWallet.privateKey, msg);
+    relayerBalance = await wallet.getBalance();
+    relayerTokenBalance = await mockToken.balanceOf(wallet.address);
+  });
+
+  it('properly construct', async () => {
+    expect(await identity.lastNonce()).to.eq(0);
+  });
+
+  describe('signing message', () => {
+    it('key exist', async () => {
+      expect(await identity.keyExist(publicKey)).to.be.true;
+    });
+
+    it('key exist', async () => {
+      expect(await identity.keyExist([])).to.be.false;
+    });
+
+    it('calculates hash', async () => {
+      const jsHash = calculateMessageHash(msg);
+      const solidityHash = await identity.calculateMessageHash(
+        msg.from,
+        msg.to,
+        msg.value,
+        msg.data,
+        msg.nonce,
+        msg.gasPrice,
+        msg.gasToken,
+        msg.gasLimit, 0);
+      expect(jsHash).to.eq(solidityHash);
+    });
+
+    it('recovers signature', async () => {
+      const recoveredAddress = await identity.getSigner(
+        msg.from,
+        msg.to,
+        msg.value,
+        msg.data,
+        msg.nonce,
+        msg.gasPrice,
+        msg.gasToken,
+        msg.gasLimit, 
+        0, 
+        signature);
+      expect(recoveredAddress).to.eq(keyAsAddress);
+    });
+  });
+
+  describe('Transfer', async () => {
+    describe('successful execution of transfer', () => {
+      it('transfers funds', async () => {
+        expect(await provider.getBalance(to)).to.eq(0);
+        await identity.executeSigned(...getExecutionArgs(msg), signature);
+        expect(await provider.getBalance(to)).to.eq(parseEther('1.0'));
+        expect(await identity.lastNonce()).to.eq(1);
+      });
+  
+      it('emits ExecutedSigned event', async () => {
+        const messageHash = calculateMessageHash(msg);
+        await expect(identity.executeSigned(...getExecutionArgs(msg), signature))
+          .to.emit(identity, 'ExecutedSigned')
+          .withArgs(messageHash, 0, true);
+      });
+
+      describe('refund', () => {
+        it('should refund in token after execute transfer ethers', async () => {
+          msg = {...transferMessage, from: identity.address, gasToken: mockToken.address};
+          signature = calculateMessageSignature(privateKey, msg);
+
+          await identity.executeSigned(...getExecutionArgs(msg), signature);
+    
+          expect(await mockToken.balanceOf(wallet.address)).to.be.above(relayerTokenBalance);
```

Comment:
this expect is super tricky, I would propose to use sth more explicit, with balance before and after